## Проект по SQL

## Описание

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.
Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

## Задачи

- Посчитайте, сколько книг вышло после 1 января 2000 года;

- Для каждой книги посчитайте количество обзоров и среднюю оценку;

- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

## Импорт библиотек

In [1]:
import pandas as pd
from sqlalchemy import create_engine

## Подключение к данным

In [2]:
db_config = {'user': 'praktikum_student', 
 'pwd': 'Sdf4$2;d-d30pp', 
 'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
 'port': 6432, 
 'db': 'data-analyst-final-project-db'} 
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

## Изучение данных

In [3]:
tables = ['books', 'authors', 'publishers', 'ratings', 'reviews']

for table in tables:
    query = f"SELECT * FROM {table}"
    data = pd.io.sql.read_sql(query, con=engine)
    
    print("\n")
    print(f"{table}")
    
    display(data)
    
    print(data.info())
    
    print("\nПроопуски:")
    print(data.isnull().sum())
    
    print("\nДубликаты:")
    print(data.duplicated().sum())



books


book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
None

Проопуски:
book_id             0
author_id           0
title               0
num_pages           0
publication_date    0
publisher_id        0
dtype: int64

Дубликаты:
0


authors


author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
None

Проопуски:
author_id    0
author       0
dtype: int64

Дубликаты:
0


publishers


publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB
None

Проопуски:
publisher_id    0
publisher       0
dtype: int64

Дубликаты:
0


ratings


rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB
None

Проопуски:
rating_id    0
book_id      0
username     0
rating       0
dtype: int64

Дубликаты:
0


reviews


review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB
None

Проопуски:
review_id    0
book_id      0
username     0
text         0
dtype: int64

Дубликаты:
0


Пропуски и дубликаты не обнаружены

## Количество книг, вышедших после 1 января 2000 года

In [4]:
query = '''
        SELECT COUNT(*) AS book_count
        FROM books
        WHERE publication_date > '2000-01-01'
        '''
result = pd.io.sql.read_sql(query, con = engine)

book_count = result['book_count'].values[0]
print("Количество книг, вышедших после 1 января 2000 года:", book_count)

Количество книг, вышедших после 1 января 2000 года: 819


## Количество обзоров и средних оценок на книги

In [5]:
query = '''
SELECT books.book_id, 
       books.title, 
       COUNT(DISTINCT review_id) AS cnt_reviews, 
       ROUND(AVG(rating), 2) AS avg_rating
       
FROM books
LEFT JOIN reviews ON books.book_id = reviews.book_id
LEFT JOIN ratings ON books.book_id = ratings.book_id

GROUP BY books.book_id, books.title
ORDER BY cnt_reviews DESC

'''
pd.io.sql.read_sql(query, con = engine)

book_id                                              title  cnt_reviews  \
0        948                            Twilight (Twilight  #1)            7   
1        963                                Water for Elephants            6   
2        734                                   The Glass Castle            6   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...            6   
4        695  The Curious Incident of the Dog in the Night-Time            6   
..       ...                                                ...          ...   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel            0   
996      808                            The Natural Way to Draw            0   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites            0   
998      221                          Essential Tales and Poems            0   
999      191  Disney's Beauty and the Beast (A Little Golden...            0   

     avg_rating  
0          3.66  
1          3.98  
2          4.21  
3          4.41  
4          4.08  
..          ...  
995        3.67  
996        3.00  
997        5.00  
998        4.00  
999        4.00  

[1000 rows x 4 columns]

Топ-3 нкиги по количеству обзоров:

- Twilight (Twilight #1) (7 rev) (3.66)

- Water for Elephants (6 rev) (3.98)

- The Glass Castle (6 rev) (4.21)

Топ-3 нкиги по рейтингу:

- Arrows of the Queen (Heralds of Valdemar #1) (2 rev)(5.00) 

- The Walking Dead Book One (The Walking Dead #1-12) (2 rev) (5.00)

- Light in August (2 rev) (5.00)

## Издательство, которое выпустило наибольшее число книг толще 50 страниц

In [6]:
query = '''
        SELECT p.publisher, COUNT(b.book_id) AS book_count
        FROM publishers p
        INNER JOIN books b ON p.publisher_id = b.publisher_id
        WHERE b.num_pages > 50
        GROUP BY p.publisher
        ORDER BY book_count DESC
        LIMIT 1
        '''
publisher_result = pd.read_sql_query(query, con=engine)

publisher = publisher_result['publisher'].values[0]
book_count = publisher_result['book_count'].values[0]
print("Издательство:", publisher)
print("Количество книг:", book_count)

Издательство: Penguin Books
Количество книг: 42


## Автор с самой высокой средней оценкой книг

In [7]:
query = '''
        SELECT a.author, AVG(r.rating) AS average_rating
        FROM authors a
        INNER JOIN (
            SELECT b.author_id, b.book_id
            FROM books b
            INNER JOIN ratings r ON b.book_id = r.book_id
            GROUP BY b.author_id, b.book_id
            HAVING COUNT(r.rating_id) >= 50
        ) AS filtered_books ON a.author_id = filtered_books.author_id
        INNER JOIN ratings r ON filtered_books.book_id = r.book_id
        GROUP BY a.author
        ORDER BY average_rating DESC
        LIMIT 1
        '''

author_result = pd.read_sql_query(query, con=engine)

author = author_result['author'].values[0]
average_rating = author_result['average_rating'].values[0]
print("Автор:", author)
print("Средняя оценка:", round(average_rating, 2))

Автор: J.K. Rowling/Mary GrandPré
Средняя оценка: 4.29


## Среднее количество обзоров от пользователей, которые поставили больше 48 оценок

In [10]:
query = '''
        SELECT AVG(count_reviews)
        FROM (
            SELECT username, COUNT(*) AS count_reviews
            FROM reviews
            WHERE username IN (
                SELECT username
                FROM ratings
                GROUP BY username
                HAVING COUNT(*) > 50)
            GROUP BY username
        ) AS review_counts
        '''
average_review_count_result = pd.read_sql_query(query, con=engine)

average_review_count = average_review_count_result['avg'].values[0]
print("Среднее количество обзоров от пользователей, которые поставили больше 48 оценок:", round(average_review_count), 2)

Среднее количество обзоров от пользователей, которые поставили больше 48 оценок: 24 2


## Вывод

В ходе выполнения исследовательской работы были изучены данные и даны ответы на поставленные вопросы

- После 1 января 2000 года вышло 819 книг

- Количество обзоров и средних оценок на книги

     Топ-3 нкиги по количеству обзоров:
     
     - Twilight (Twilight #1) (7 rev) (3.66 rat)
     
     - Water for Elephants (6 rev) (3.98 rat)
     
     - The Glass Castle (6 rev) (4.21 rat)
     
    Топ-3 нкиги по рейтингу:
    
     - Arrows of the Queen (Heralds of Valdemar #1) (2 rev) (5.00 rat)
     
     - The Walking Dead Book One (The Walking Dead #1-12) (2 rev) (5.00 rat)
     
     - Light in August (2 rev) (5.00 rat)

- Penguin Books, выпустило наибольшее число книг толще 50 страниц (42)

- J.K. Rowling/Mary GrandPré, автор с самой высокой средней оценкой книг (4.29 rat)

- Среднее количество обзоров от пользователей, которые поставили больше 48 оценок равно 24 rev